core/flyout_horizontal.js

```diff
@@ -0,0 +1,456 @@
+ * @private
+ */
+Blockly.HorizontalFlyout.prototype.layout_ = function(contents, gaps) {
+  this.workspace_.scale = this.targetWorkspace_.scale;
+  var margin = this.MARGIN;
+  var cursorX = this.RTL ? margin : margin + Blockly.BlockSvg.TAB_WIDTH;
+  var cursorY = margin;
+  if (this.RTL) {
+    contents = contents.reverse();
+  }
+
+  for (var i = 0, item; item = contents[i]; i++) {
+    if (item.type == 'block') {
+      var block = item.block;
+      var allBlocks = block.getDescendants();
+      for (var j = 0, child; child = allBlocks[j]; j++) {
+        // Mark blocks as being inside a flyout.  This is used to detect and
+        // prevent the closure of the flyout if the user right-clicks on such a
+        // block.
+        child.isInFlyout = true;
+      }
+      block.render();
+      var root = block.getSvgRoot();
+      var blockHW = block.getHeightWidth();
+
+      // Figure out where to place the block.
+      var tab = block.outputConnection ? Blockly.BlockSvg.TAB_WIDTH : 0;
+      if (this.RTL) {
+        var moveX = cursorX + blockHW.width;
+      } else {
+        var moveX = cursorX + tab;
+      }
+      block.moveBy(moveX, cursorY);
+
+      var rect = this.createRect_(block, moveX, cursorY, blockHW, i);
+      cursorX += (blockHW.width + gaps[i]);
+
+      this.addBlockListeners_(root, block, rect);
+    } else if (item.type == 'button') {
+      this.initFlyoutButton_(item.button, cursorX, cursorY);
+      cursorX += (item.button.width + gaps[i]);
+    }
+  }
+};
+
+/**
+ * Determine if a drag delta is toward the workspace, based on the position
+ * and orientation of the flyout. This is used in determineDragIntention_ to
+ * determine if a new block should be created or if the flyout should scroll.
+ * @param {!goog.math.Coordinate} currentDragDeltaXY How far the pointer has
+ *     moved from the position at mouse down, in pixel units.
+ * @return {boolean} true if the drag is toward the workspace.
+ * @package
+ */
+Blockly.HorizontalFlyout.prototype.isDragTowardWorkspace = function(
+    currentDragDeltaXY) {
+  var dx = currentDragDeltaXY.x;
+  var dy = currentDragDeltaXY.y;
+  // Direction goes from -180 to 180, with 0 toward the right and 90 on top.
+  var dragDirection = Math.atan2(dy, dx) / Math.PI * 180;
+
+  var range = this.dragAngleRange_;
+  if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_TOP) {
+    // Horizontal at top.
+    if (dragDirection < 90 + range && dragDirection > 90 - range) {
+      return true;
+    }
+  } else {
+    // Horizontal at bottom.
+    if (dragDirection > -90 - range && dragDirection < -90 + range) {
+      return true;
+    }
+  }
+  return false;
+};
+
+/**
+ * Copy a block from the flyout to the workspace and position it correctly.
+ * @param {!Blockly.Block} originBlock The flyout block to copy..
+ * @return {!Blockly.Block} The new block in the main workspace.
+ * @private
+ */
+Blockly.HorizontalFlyout.prototype.placeNewBlock_ = function(originBlock) {
```

Comment:
I think horizontal & vertical versions of this are almost identical.  It'd be nice if we could break this up and push a lot of this back up to flyout base eventually.   y'know, when we have lots of free time :)